### PR TITLE
Optimized recursive_bubble_sort

### PR DESCRIPTION
**Pull request by shellhub**
### **Describe your change:**

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by shellhub** [comment]
@cclauss Please review. Thanks in advance :)

File: sorts/recursive_bubble_sort.py [hunk 1]
```diff
@@ -1,41 +1,41 @@
-def bubble_sort(list1):
+def bubble_sort(list_data: list, length: int) -> list:
```
Comment (cclauss) [inline sorts/recursive_bubble_sort.py:1]:
```suggestion
def bubble_sort(list_data: list, length: int=0) -> list:
    length = length or len(list_data)
```

File: sorts/recursive_bubble_sort.py [hunk 2]
```diff
@@ -1,41 +1,41 @@
...
-    >>> bubble_sort([-2, -45, -5])
+    >>> bubble_sort([-2, -45, -5], 3)
     [-45, -5, -2]
 
-    >>> bubble_sort([-23, 0, 6, -4, 34])
+    >>> bubble_sort([-23, 0, 6, -4, 34], 5)
     [-23, -4, 0, 6, 34]
 
-    >>> bubble_sort([-23, 0, 6, -4, 34]) == sorted([-23, 0, 6, -4, 34])
+    >>> bubble_sort([-23, 0, 6, -4, 34], 6) == sorted([-23, 0, 6, -4, 34])
```
Comment (cclauss) [inline sorts/recursive_bubble_sort.py:20]: Why does this work?  len(list_data) is 5 but length is 6

File: sorts/recursive_bubble_sort.py (hunk 1 above)
Comment (shellhub) [reply sorts/recursive_bubble_sort.py:1]: Why need change like this ?

File: sorts/recursive_bubble_sort.py (hunk 2 above)
Comment (shellhub) [inline sorts/recursive_bubble_sort.py:20]:
```suggestion
    >>> bubble_sort([-23, 0, 6, -4, 34], 5) == sorted([-23, 0, 6, -4, 34])
```

File: sorts/recursive_bubble_sort.py (hunk 1 above)
Comment (cclauss) [reply sorts/recursive_bubble_sort.py:1]: If I want to sort a list, why do I need to tell you how long it is in a language like Python that has len() builtin?

**Review by shellhub — COMMENTED** [review]
@cclauss fixed

File: sorts/recursive_bubble_sort.py (hunk 1 above)
Comment (shellhub) [reply sorts/recursive_bubble_sort.py:1]: Did you mean optional argument ? can you edit like this?
``` python
def bubble_sort(list_data: list, length: int = 0) -> list:
```

**Comment by cclauss** [comment]
https://travis-ci.com/github/TheAlgorithms/Python/builds/183314106#L255

File: sorts/recursive_bubble_sort.py [hunk 3]
```diff
@@ -1,41 +1,42 @@
...
-    >>> bubble_sort([-23, 0, 6, -4, 34]) == sorted([-23, 0, 6, -4, 34])
+    >>> bubble_sort([-23, 0, 6, -4, 34], 5) == sorted([-23, 0, 6, -4, 34])
     True
 
-    >>> bubble_sort(['z','a','y','b','x','c'])
+    >>> bubble_sort(['z','a','y','b','x','c'], 6)
     ['a', 'b', 'c', 'x', 'y', 'z']
 
+    >>> bubble_sort([1.1, 3.3, 5.5, 7.7, 2.2, 4.4], 6) == sorted([1.1, 3.3, 5.5, 7.7, 2.2, 4.4])
+    True
```
Comment (shellhub) [inline sorts/recursive_bubble_sort.py:28]:
```suggestion
    >>> bubble_sort([1.1, 3.3, 5.5, 7.7, 2.2, 4.4, 6.6])
    [1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7]
```

**Review by shellhub — COMMENTED** [review]
fix build error

**Comment by shellhub** [comment]
@cclauss why Travic CI occur error?

**Comment by cclauss** [comment]
Probably trailing whitespace deleted in doctests.

**Comment by shellhub** [comment]
How to solve？😭

**Comment by cclauss** [comment]
See which files are failing in Travis CI and then undo the changes to those files.

**Comment by shellhub** [comment]
Yes. I found why build error was occur. Because whitespace deleted in doctests. Should I update origin source code and commit again? have you any better suggestion? @cclauss 
image

**Comment by spamegg1** [comment]
The build log is very difficult to understand.
It's showing all kinds of errors from OTHER files.
What is going on?
Also it says above "60 files changed" in shellhub:dev branch, was this your intention @shellhub ?
I suppose this was caused by the forced push that applied psf/black to everything?
It might be best to just start over with 1 file changed.

**Comment by cclauss** [comment]
This was more than just running black.  In the cases of the failing doctests whitespace has been removed from the end of lines which the doctest needs to see for an exact match.  This is one of the reasons why black will not suppress trailing whitespace that appears inside of triple quoted strings.  There are three files that fail the Travis doctests.  If the changes to those three files are reverted then the tests will pass again.

File: traversals/binary_tree_traversals.py [hunk 4]
```diff
@@ -241,7 +241,7 @@ def post_order_iter(node: TreeNode) -> None:
     >>> tree_node2.left, tree_node2.right = tree_node4 , tree_node5
     >>> tree_node3.left, tree_node3.right = tree_node6 , tree_node7
     >>> post_order_iter(root)
-    4 5 2 6 7 3 1 
+    4 5 2 6 7 3 1
```
Comment (shellhub) [inline traversals/binary_tree_traversals.py:244]:
```suggestion
    4 5 2 6 7 3 1 
```

**Review by shellhub — COMMENTED** [review]
Added whitespace

**Comment by cclauss** [comment]
My sense is that you add the trailing whitespace and then the autoblack GitHub Action undoes it so the tests fail.

**Comment by shellhub** [comment]
@cclauss All checks have passed. Can you review again. Thanks :)

**Comment by cclauss** [comment]
I am completely lost by the change to bubble sort!  We now require the user to provide `length` but our algorithm never uses that variable.

**Comment by shellhub** [comment]
> I am completely lost by the change to bubble sort! We now require the user to provide `length` but our algorithm never uses that variable.

Not exactly. `length` is optional to user. And we used `length` in our recursion code.
``` python
return list_data if not swapped else bubble_sort(list_data, length - 1)
```

**Comment by shellhub** [comment]
Get a maximum value for each recursive call. Then recursion sort left elements (`length - 1` elements)

**Comment by cclauss** [comment]
Please look at it again.  Passing a number around but not actually using it (outside of decrementing it and passing it back to yourself) is a waste of cycles.

**Comment by shellhub** [comment]
@cclauss you are right. I forget make changes that you mentioned before. I make changes like this. Please review again. Thanks :)
``` python
    length = length or len(list_data)
    swapped = False
    for i in range(length - 1):
        if list_data[i] > list_data[i + 1]:
            list_data[i], list_data[i + 1] = list_data[i + 1], list_data[i]
            swapped = True

    return list_data if not swapped else bubble_sort(list_data, length - 1)
```

**Comment by shellhub** [comment]
I changed `for i in range(len(list_data) - 1):` to `for i in range(length - 1):`

**Comment by cclauss** [comment]
OK...  Here is what we are going to do.  I will land this now.  From now on, no PRs that touch so many files.

Please create a new PR for bubble sort that has both the old algorithm and the "_oprimized_" version in a single file with a timeit (or similar) benchmark that measures the relative performance of each.

**Review by cclauss — APPROVED** [review]
LGTM

**Pull request merged**